guardrails-ai/guardrails
# In Code Hub Installs

Added:
- In code hub installs using `from guardrails import install`
- Slightly more test coverage of install utility functions

Changed:
- Made installs slightly faster by avoiding subprocess to get site packages locations 
- Moved installation utilities and logic to main `guardrails.hub` module as opposed to `cli` module (Phase 1)

Has been tested in Jupyter Notebooks 📖  (which is one aspect of what motivated this feature)

Usage:

```python
from guardrails import Guard, OnFailAction, install


install("hub://guardrails/contains_string") # default is quiet=True

from guardrails.hub import ContainsString

guard = Guard().use(ContainsString, substring="s", on_fail="exception")

try:
    response = guard.validate("pass")   # Guardrail passes
    print("Expected: Validation passes")
except Exception as e:
    print("Unexpected: ",e)

try:
    response = guard.validate("fail")  # Guardrail fails
    print("Unexpected: Validation passes")
except Exception as e:
    print("Expected: ", e)
```

```python
from guardrails import install, Guard, OnFailAction


module = install("hub://guardrails/regex_match", quiet=False, install_local_models=True)

RegexMatch = module.RegexMatch

guard = Guard().use(
    RegexMatch, regex="\(?\d{3}\)?-? *\d{3}-? *-?\d{4}", on_fail=OnFailAction.EXCEPTION
)

try:
    guard.validate("123-456-7890")  # Guardrail passes
    print("Expected: Validation passes")
except Exception as e:
    print("Unexpected: ",e)

try:
    guard.validate("1234-789-0000")  # Guardrail fails
    print("Unexpected: Validation passes")
except Exception as e:
    print("Expected: ", e)
```

Follow up improvements:
- Update `create` cli to use install function
- Remove `[notice]` logs which breaks up installation progress bar
- Remove duplicated functionality (used in `create`)
